Fixed #32316 -- Deferred accessing __file__.

wkschwartz opened the pull request:
This fixes [#32316]. As I explained there, my implementation may warrant some discussion. I'm submitting this PR broken into separate commits to facilitate that discussion and cherry-picking, if needed. I am aware it will need to be squashed before being merged. The commits that are most likely to warrant discussion are 07f46b7b26a1a74fa531de834bee240d576b7f85, which modifies `django.views.debug.ExceptionReporter`, and 24aa80ba413eeaaec0f3c2cab2f028e0dcbaf99b, which modifies `django.contrib.auth.password_validation.CommonPasswordValidator`.

[#32316]: https://code.djangoproject.com/ticket/32316

orf commented on django/contrib/auth/password_validation.py:
```diff
@@ -167,9 +167,20 @@ class CommonPasswordValidator:
     """
-    DEFAULT_PASSWORD_LIST_PATH = Path(__file__).resolve().parent / 'common-passwords.txt.gz'
+
+    @property
```
Lets use `django.utils.functional.cached_property` here instead?

orf commented on django/utils/version.py:
```diff
@@ -76,6 +76,8 @@ def get_git_changeset():
     This value isn't guaranteed to be unique, but collisions are very unlikely,
     so it's sufficient for generating the development version numbers.
     """
+    if '__file__' not in globals():
```
Can we add a small comment explaining this, for future reference?

wkschwartz replied:
GitHub isn't marking this as outdated now, but I'm hoping that 7846f42 is responsive to your request.

wkschwartz commented:
@orf is there anything else you'd like from this PR? I think my last two commits addressed your previous comments, and the failing tests appear unrelated to this PR. If you're happy with it, would you like me to `git rebase master && git push --force`?

shreyasravi reviewed (approved):
Looks good to me.

carltongibson commented on django/utils/version.py:
```diff
@@ -76,6 +76,11 @@ def get_git_changeset():
     This value isn't guaranteed to be unique, but collisions are very unlikely,
     so it's sufficient for generating the development version numbers.
     """
+    # If __file__ is undefined in this module, we won't be able to find the
+    # repo. Besides, we're likely running inside a frozen module, which isn't
+    # necessarily frozen from the repo's working tree in the first place.
```
See [the coding style guide here](https://docs.djangoproject.com/en/3.1/internals/contributing/writing-code/coding-style/#python-style). Avoid 'we' in comments. 

Maybe 

```suggestion
    # Repo can't be found if __file__ is undefined, e.g. in a frozen module.
```

carltongibson commented on django/views/debug.py:
```diff
@@ -259,6 +259,11 @@ def __init__(self, request, exc_type, exc_value, tb, is_email=False):
         self.template_info = getattr(self.exc_value, 'template_debug', None)
         self.template_does_not_exist = False
         self.postmortem = None
+        # Avoid using __file__ at module level to be kind to frozen Python environments
```
Could this comment be a docstring on the helper function? 

I'd probably rephrase slightly.

wkschwartz replied:
Is 93124234f2febfc48e92a910243f42f4ebf5d72a responsive to this request?

carltongibson commented on tests/version/tests.py:
```diff
@@ -12,6 +12,18 @@ def test_development(self):
 
+    def test_development_no_file(self):
+        import django.utils.version
+        if not hasattr(django.utils.version, '__file__'):
+            return  # test_development() already accomplished this test's goal
```
As we run the test suite, does this ever evaluate True? (i.e. Do we need this branch?)

wkschwartz replied:
I think [eventually](https://code.djangoproject.com/ticket/32316#comment:4) I may want to [propose](https://code.djangoproject.com/ticket/32316#comment:6) that Django run tests (or at least _be able to be tested_) in a frozen environment. I agree that for now this branch won't get used. But it seems to me to that it is of at most de minimis cost to include this branch for technical correctness—the whole point of this patch is that [`__file__`](https://docs.python.org/3/reference/import.html#__file__) may not exist—to support testing in a frozen environment in the future.

That said, I won't die on this hill. If you want the branch gone, I'll take it out.

felixxm replied:
Use `skipIf()`.

wkschwartz replied:
I added `skipIf()` in 17c18e126cc8572b20424268d59c73223e299263.

But then I realized that what I cared about here was the test passing when `__file__` is missing, not avoiding duplication with `test_development()`. 40059b2b988fba0417a3c768eb87e29850f19504 simplifies this a little (it's 3 lines shorter) by allowing `test_development_no_file` to run when `__file__` is missing.

It [doesn't matter yet](https://github.com/django/django/pull/13841#discussion_r570230164), but I manually tested that both commits work in the absence of `__file__` with the following very stupid patch:

```diff
diff --git a/tests/runtests.py b/tests/runtests.py
index afdf68c21b..6e13ec3ad5 100755
--- a/tests/runtests.py
+++ b/tests/runtests.py
@@ -29,6 +29,9 @@ else:
     )
     from django.utils.log import DEFAULT_LOGGING
 
+import django.utils.version
+del django.utils.version.__file__
+
 try:
     import MySQLdb
 except ImportError:
```

and running `./runtests.py version` from the `tests` directory.

felixxm replied:
> But then I realized that what I cared about here was the test passing when `__file__` is missing, not avoiding duplication with `test_development()`.

`test_development()` already checks this scenario. I don't see much value in duplicated tests.

carltongibson commented on django/utils/module_loading.py:
```diff
@@ -86,7 +86,7 @@ def module_dir(module):
     over several directories.
     """
-    # Convert to list because _NamespacePath does not support indexing.
+    # Convert to list because __path__ is only required to be Iterable[str]
```
Maybe: 

```suggestion
    # Convert to list because __path__ may not support indexing.
```

carltongibson commented on django/views/debug.py:
```diff
@@ -26,7 +26,9 @@
 
-CURRENT_DIR = Path(__file__).parent
+
+def _builtin_template_path(name):
```
Not sure this needs the `_` 🤷

wkschwartz replied:
I didn't want to introduce any public API in a "cleanup/optimization" PR.

felixxm replied:
This hook is used only in `ExceptionReporter`, so I would change this to a `ExceptionReporter` method.

> I didn't want to introduce any public API in a "cleanup/optimization" PR.

API is not considered public if it's not documented.

wkschwartz replied:
> This hook is used only in `ExceptionReporter`

It's also used in `technical_404_response` and `default_urlconf`. I'll hold off making it an `ExceptionReporter` method unless you confirm you still want me to do that.

> API is not considered public if it's not documented.

I'll get rid of the underscore.

wkschwartz replied:
> I'll get rid of the underscore.

Done: 9adc5e75db5bfa2b91956424b0b349f7f178787e

carltongibson commented on django/apps/config.py:
```diff
@@ -73,8 +73,7 @@ def _path_from_module(self, module):
         # various cases.
-        # Convert paths to list because Python's _NamespacePath doesn't support
-        # indexing.
+        # Convert to list because __path__ is only required to be Iterable[str]
```
Same comment as other occurrence.

carltongibson reviewed (commented):
Hi @wkschwartz. Thanks for this. 

A few stylistic comments… Worth going over the [Patch review checklist](https://docs.djangoproject.com/en/3.1/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). Can you rebase, squash to a single commit and use the required [commit message format](https://docs.djangoproject.com/en/3.1/internals/contributing/committing-code/#committing-guidelines). 

Ref `cached_property` I think we'll be good to prefer the Python version from dropping 3.7 shortly. (I need to double check on our plans to deprecate that… 🤔)

I'm a little bit Mmmm overall, but I think OK, given your goal to use frozen envs it should be acceptable. Thanks for the effort!

wkschwartz commented on django/contrib/auth/password_validation.py:
```diff
@@ -8,7 +8,7 @@
     FieldDoesNotExist, ImproperlyConfigured, ValidationError,
 )
-from django.utils.functional import lazy
+from django.utils.functional import cached_property, lazy
```
Does your request mean there's no chance this PR will be back ported to 3.2?

jacobtylerwalls replied:
Right, we don't backport optimizations. Just fixes to issues that would have been release blockers in the current version, as I understand it.

wkschwartz commented:
@carltongibson I hope I have addressed everything in your review. Would you mind taking another look? Thanks for the feedback!

jacobtylerwalls commented:
Hi @wkschwartz if you've addressed all feedback to date, then be sure to untick Patch Needs Improvement on the ticket.

wkschwartz commented:
Thanks @jacobtylerwalls. Done.

orf reviewed (approved):
LGTM

felixxm commented on django/contrib/auth/password_validation.py:
```diff
@@ -167,9 +167,14 @@ class CommonPasswordValidator:
     """
-    DEFAULT_PASSWORD_LIST_PATH = Path(__file__).resolve().parent / 'common-passwords.txt.gz'
+
+    @functools.cached_property
```
Please use `cached_property` from `django.utils.functional`.

wkschwartz replied:
I will do this.

However, please be aware that this line of code has become something of a football among reviewers. https://github.com/django/django/pull/13841#discussion_r551560566 asked for `django.utils.functional`. Then https://github.com/django/django/pull/13841#discussion_r570091753 asked for `functools`. And now we're back to `django.utils.functional`. I personally don't care—as I said, I'll make the requested change.

wkschwartz replied:
Done: 6fe617988edc1326511ca0840b65f9c2e9a2a6fb

felixxm replied:
ticket-30949 is currently not accepted due to some performance issues, so we should use `cached_property` from `django.utils.functional`. We can swap all uses of `cached_property` when implementing ticket-30949.

felixxm commented on django/views/debug.py:
```diff
@@ -259,6 +264,10 @@ def __init__(self, request, exc_type, exc_value, tb, is_email=False):
         self.template_info = getattr(self.exc_value, 'template_debug', None)
         self.template_does_not_exist = False
         self.postmortem = None
+        if not hasattr(self, 'html_template_path'):
+            self.html_template_path = _builtin_template_path('technical_500.html')
+        if not hasattr(self, 'text_template_path'):
+            self.text_template_path = _builtin_template_path('technical_500.txt')
```
I'm not convinced about this change is a bit backward incompatible, e.g.
```python
>>> from django.views.debug import ExceptionReporter
>>> ExceptionReporter.html_template_path
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'ExceptionReporter' has no attribute 'html_template_path'
```
Can we use a `@classproperty` (with `@lru_cache(maxsize=1)`) to keep the current behavior? :thinking:

wkschwartz replied:
Ah yes, sorry about this one. I knew it wouldn't be perfectly backward compatible when I wrote it. At the time, it was before the 3.2 feature freeze, and `ExceptionReporter`'s template paths were still new (#13531). Now that this patch is targeting 4.0, I can see we need to be more careful.

In 144b0ba8ee8b8d24cea6b766dd12ca6dcc67fcfc I added your [test](https://github.com/django/django/pull/13841/files#diff-57d9651af19f1a34c685f3f561efc7fc25fabb143729640d954f1ed929770737R929-R934) and implemented `ExceptionReporter.{html,text}_template_path` with `@classproperty` and `@lru_cache(maxsize=1)` as you suggested.

However, then I realized that creating a subclass `K` of `ExceptionReporter` and accessing `K.text_template_path` would evict `ExceptionReporter.text_template_path` from the `functools.lru_cache`, thereby defeating the purpose of using the cache in the first place. (As I understand it, the intended use case of `ExceptionReporter.text_template_path` is for easy subclassing of `ExceptionReporter`.) In aa93375963d952c3485511955cf6f536f7fbbc8e, I [amended the test](https://github.com/django/django/pull/13841/files#diff-57d9651af19f1a34c685f3f561efc7fc25fabb143729640d954f1ed929770737R936-R941) and re-implemented the caching manually so that `K.text_template_path is
ExceptionReporter.text_template_path` holds. 

If you prefer the `@lru_cache(maxsize=1)` implementation, it's easy enough to revert the latter commit.

felixxm reviewed (commented):
@wkschwartz Thanks :+1: I left comments.

wkschwartz commented:
@felixxm I hope I have addressed everything in your review (per https://github.com/django/django/pull/13841#issuecomment-786903448 I've [unset "Patch needs improvement"](https://code.djangoproject.com/ticket/32316#comment:12)). Would you mind taking another look? Thanks for the feedback!

felixxm commented on django/apps/config.py:
```diff
@@ -73,8 +73,7 @@ def _path_from_module(self, module):
         # various cases.
-        # Convert paths to list because Python's _NamespacePath doesn't support
-        # indexing.
+        # Convert to list because __path__ may not support indexing.
```
IMO this change is rather related with ticket-32355, because `_NamespacePath` supports indexing in Python 3.8+, see https://github.com/python/cpython/commit/ab9b31f94737895f0121f26ba3ad718ebbc24fe1. I added a separate PR #14192.

wkschwartz replied:
Got it. I rebased on top of main, so this line isn't part of this PR anymore.

wkschwartz commented:
@felixxm It looks like you did a follow up review but didn't comment with anything I should change. Does this PR need anything else or is it ready for check in?

carltongibson reviewed (commented):
On the current suggestion this test passes, which is desired right: 

```diff
diff --git a/tests/view_tests/tests/test_debug.py b/tests/view_tests/tests/test_debug.py
index bcc00928ac..d7c798b8d3 100644
--- a/tests/view_tests/tests/test_debug.py
+++ b/tests/view_tests/tests/test_debug.py
@@ -939,6 +939,13 @@ class ExceptionReporterTests(SimpleTestCase):
         self.assertIs(K.html_template_path, ExceptionReporter.html_template_path)
         self.assertIs(K.text_template_path, ExceptionReporter.text_template_path)
 
+        # classproperty descriptor can be overridden.
+        class J(ExceptionReporter):
+            html_template_path = Path(__file__)
+            text_template_path = Path(__file__)
+        self.assertEqual(J.html_template_path, Path(__file__))
+        self.assertEqual(J.text_template_path, Path(__file__))
+
 
 class PlainTextReportTests(SimpleTestCase):
     rf = RequestFactory()
```

This is already covered by `TemplateOverrideExceptionReporter`. 

Are there any issues with this approach? 

We need to consider whether we should incorporate this into 3.2 now, before final release, rather than immediately introduce a change. 🤔 

This would require updating the the updating the release note in 3.2.txt, and the docs in error-reporting.txt. 

* What's the rst/sphinx for a `@classproperty`? 😬 — But examples using a string would probably be helpful.

For this last, something like... 

```
    .. method:: html_template_path

        .. versionadded:: 3.2
        
        A ``@classproperty`` returning a :class:`pathlib.Path` representing the
        absolute filesystem path to a template for rendering the HTML
        representation of the exception. Defaults to the Django provided
        template.

        Subclasses may usually override the ``@classproperty`` descriptor with
        a :class:`pathlib.Path` instance directly.
```

Is ` .. method::` the right one to use? 🙂

carltongibson commented:
@wkschwartz — Would you be up for changing these ExceptionReporter attributes to (plain) properties? We can backport that to 3.2 now, and then it's cleaner here.

carltongibson commented:
To answer my own question, we doc properties as attributes: 

https://github.com/django/django/blob/038940cf5525c41464a1b9e9ba3801042320b0cc/docs/ref/forms/api.txt#L106

wkschwartz commented:
> @wkschwartz — Would you be up for changing these ExceptionReporter attributes to (plain) properties? We can backport that to 3.2 now, and then it's cleaner here. 

Yep. I'll submit a PR for 3.2 later today. Once we've nailed down the implementation, I'll update this PR (which IIUC is for 4.0) based on what y'all accept for 3.2.

carltongibson commented:
@wkschwartz Please make it to master. I/Mariusz will handle the backport. Thanks. 👍

wkschwartz commented:
#14203 addresses @carltongibson's request.

wkschwartz commented:
@carltongibson @felixxm I rebased onto `main` to incorporate 7248afe12f40361870388ecdd7e0038eb0d58e47. I got rid of caching in `ExceptionReporter.{html,text}_template_path` because caching isn't pertinent to this PR. Happy to add it back here or in a separate PR if you'd prefer. I also cleaned up the commit message a bit. I've re-requested your reviews as I believe the PR is now ready to go.

felixxm commented on tests/version/tests.py:
```diff
@@ -12,6 +12,20 @@ def test_development(self):
+        try:
+            if version_file is not MISSING:
+                del django.utils.version.__file__
+            ver_string = get_version((1, 4, 0, 'alpha', 0))
```
We need to clear cache before calling `get_version()`.

felixxm reviewed (commented):
@wkschwartz Thanks :+1:

I pushed final edits.

wkschwartz reviewed (commented):
Thanks for wrapping this up, @felixxm! Just a couple minor suggestions—none of which are critical—on the changes you made.

felixxm commented on tests/version/tests.py:
```diff
@@ -1,17 +1,37 @@
+        'e.g. in a frozen environments'
+    )
+    def test_development_no_file(self):
+        get_git_changeset.cache_clear()
```
I wanted to keep them isolated, without calling `cache_clear()` in both one of them will always be no-op.

felixxm commented:
@wkschwartz Thanks :+1: I pushed edits.
